src/java_bytecode/generate_java_generic_method.cpp

```diff
@@ -0,0 +1,296 @@
+  specialized_method.name=
+    id2string(specialized_method.name)+signature_decoration;
+
+  for(code_typet::parametert &parameter : specialized_code.parameters())
+  {
+    // Update the symbol name of the parameter to match the newly decorated
+    // name of the specialized method.
+    decorate_identifier(parameter, ID_C_identifier, original_name,
+      signature_decoration);
+
+    typet &parameter_type=parameter.type();
+
+    if(is_java_generic_parameter(parameter_type))
+    {
+      instantiate_java_generic_parameter(
+        to_java_generic_parameter(parameter_type),
+        concrete_type_map);
+    }
+    else if(is_java_generic_type(parameter_type))
+    {
+      instantiate_java_generic_type(
+        to_java_generic_type(parameter_type),
+        concrete_type_map);
+    }
+  }
+
+  // Updated the body of the specialized method so that all references to
+  // symbols in the method body refer to the specialized method and not the
+  // generic method.
+  // FIXME: Probably also need to update references to fields to point to the
+  // FIXME: specialized class fields, rather than the generic class fields?
+  auto body_iterator=specialized_method.value.depth_begin();
+  const auto body_end=specialized_method.value.depth_end();
+
+  while(body_iterator!=body_end)
+  {
+    if(body_iterator->id()==ID_symbol)
+    {
+      exprt &symbol_expr=body_iterator.mutate();
+      typet &symbol_type=symbol_expr.type();
+      if(is_java_generic_parameter(symbol_type))
+      {
+        instantiate_java_generic_parameter(
+          to_java_generic_parameter(symbol_type),
+          concrete_type_map);
+      }
+      else if(is_java_generic_type(symbol_type))
+      {
+        instantiate_java_generic_type(
+          to_java_generic_type(symbol_type),
+          concrete_type_map);
+      }
+      decorate_identifier(symbol_expr, ID_identifier, original_name,
+        signature_decoration);
+    }
+
+    ++body_iterator;
+  }
+
+  if(symbol_table.add(specialized_method))
+  {
+    // We should never hit this because the first thing
+    // we do before starting to specialize the method
+    // is check that the proposed symbol name doesn't
+    // already exist.
+    throw "specialized method symbol already exists";
+  }
+
+  return *symbol_table.lookup(specialized_method.name);
+}
+
+
+/// Given a mapping from generic type variables to concrete types, generate
+/// a string suitable for decorating symbol names with.
+/// \param concrete_types The mapping from type variables to concrete types
+/// \return A string suitable for inserting into a java signature
+///
+/// As an example, if a mapping such as this (pseudo code...):
+/// [ {T,java.lang.Integer}, {U,java.lang.Double} ] is passed in, the return
+/// string would look like "<java.lang.Integer,java.lang,Double>"
+const std::string generate_java_generic_methodt::instantiation_decoration
+  (type_variable_instantiationst concrete_types) const
```

Comment:
Nit: break after the bracket:
```C++
const std::string generate_java_generic_methodt::instantiation_decoration(
  type_variable_instantiationst concrete_types) const
```

Reply:
Ideally I suppose we should pull this out into some kind of generic_utils since this duplicates the code in @NlightNFotis's work

Reply:
Yeah, this probably is common so worth factoring between @NlightNFotis' and my own work.